Source/WebCore/editing/Editing.cpp

```diff
@@ -1499,6 +1502,83 @@ SimpleRange adjustToVisuallyContiguousRange(const SimpleRange& range)
     }, SelectionExtentMovement::Closest).value_or(range);
 }
 
+EnclosingLayerInfomation computeEnclosingLayer(bool isContentEditable, const SimpleRange& range)
+{
+    VisibleSelection selection = VisibleSelection(range);
+
+    if (selection.isNoneOrOrphaned())
+        return { nullptr, nullptr, nullptr, { } };
+
+    if (!isContentEditable && selection.isCaret())
+        return { nullptr, nullptr, nullptr, { } };
+
+    auto findEnclosingLayer = [](const Position& position) -> RenderLayer* {
+        RefPtr container = position.containerNode();
+        if (!container)
+            return nullptr;
+
+        CheckedPtr renderer = container->renderer();
+        if (!renderer)
+            return nullptr;
+
+        return renderer->enclosingLayer();
+    };
+
+    auto [startLayer, endLayer] = [&] -> std::pair<CheckedPtr<RenderLayer>, CheckedPtr<RenderLayer>> {
+        if (RefPtr container = selection.start().containerNode(); container && ImageOverlay::isInsideOverlay(*container)) {
+            RefPtr host = container->shadowHost();
+            if (!host) {
+                ASSERT_NOT_REACHED();
+                return { };
+            }
+
+            CheckedPtr renderer = host->renderer();
+            if (!renderer)
+                return { };
+
+            CheckedPtr enclosingLayer = renderer->enclosingLayer();
+            return { enclosingLayer, enclosingLayer };
+        }
+
+        return { findEnclosingLayer(selection.start()), findEnclosingLayer(selection.end()) };
+    }();
+
+    if (!startLayer)
+        return { nullptr, nullptr, nullptr, { } };
+
+    if (!endLayer)
+        return { nullptr, nullptr, nullptr, { } };
```

Comment:
This might also look a bit cleaner:

```suggestion
        return { };

    if (!endLayer)
        return { };
```